src/mercury_engine_data_structures/formats/audio_common.py

```diff
@@ -0,0 +1,147 @@
+    Tell,
+)
+from construct.expr import Path
+from construct.lib.containers import Container, ListContainer
+
+from mercury_engine_data_structures.common_types import VersionAdapter, make_dict
+from mercury_engine_data_structures.construct_extensions.alignment import AlignTo
+
+# An int32 offset (or size) that can be rebuilt later in another struct using RebuildOffset().
+# stores the offset without needing extra Tells or private offsets cluttering the struct.
+RebuildableOffset = Struct(
+    "_offset" / Tell,
+    "value" / Int32ul,
+)
+
+def RebuildOffset(target_offset: Path, starting_from: int | Path = 0) -> Struct:
+    """
+    Returns a struct that will rebuild the targeted RebuildableOffset with the current stream's position.
+
+    :param target_offset: a construct path to be rebuilt
+    :param starting_from: an optional construct path or lambda to rebuild a relative offset or size a structure
+    """
+    return Struct(
+        "cur_offset" / Tell,
+        "updated" / Pointer(target_offset._offset, Rebuild(Int32ul, construct.this.cur_offset - starting_from)),
+    )
+
+def Header(magic: bytes, version: int | str | tuple[int, int, int], sections: dict[str, int]) -> Struct:
+    """
+    Returns an AAL header for the given magic, version number, and sections. Expects to be a little-endian file.
+
+    :param magic: a bytestring containing the fourcc (FSTM, FSAR, etc)
+    :param version: a string or tuple representing the version number for a VersionAdapter
+    :param sections: a dictionary mapping data block fourcc's (STRG, INFO, FILE, etc) to their section flags
+    """
+    return Struct(
+        "_magic" / Const(magic),
+        Const(0xFEFF, Int16ul), # byte-order mark, always LE
+        "_header_size_offset" / Tell,
+        "_header_size" / Int16ul,
+        Const(0, Byte),
+        "version" / VersionAdapter(version, (Byte, Byte, Byte)),
+        "size" / RebuildableOffset,
+        "sections" / make_dict(
+            value=Struct(
+                "offset" / RebuildableOffset,
+                "size" / RebuildableOffset,
+            ),
+            key = Enum(Int32ul, **sections)
+        ),
+        AlignTo(0x20),
+        "_end_offset" / Tell,
+        Pointer(
+            construct.this._header_size_offset,
+            Rebuild(Int16ul, construct.this._end_offset - construct.this._header_size_offset + 6)
+        )
+    )
+
+def Block(name: str, subcon: Struct):
+    """
+    Uses the subcon to parse the data block, and updates the offset and size in the header.
+
+    :param name: the section fourcc (used to find header context)
+    :param subcon: the subcon to parse the data block with
+    """
+    return FocusedSeq(
+        "blk",
+        "_start" / Tell,
+        RebuildOffset(construct.this._root.header.sections[name]["offset"]),
+        "blk" / subcon,
+        RebuildOffset(construct.this._root.header.sections[name]["size"], construct.this._._start),
+    )
+
+class ReferenceTableOfOffsets(Adapter):
```

Comment:
There's several similar variants of this (ReferenceTable where the table is directly pointing to data, ReferenceTable where the table contains data size as well as offset, etc). not currently sure if i'll rename it and make it more flexible or if i'll make several ReferenceTable classes.